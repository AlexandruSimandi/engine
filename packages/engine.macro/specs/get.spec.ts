import pluginTester from "babel-plugin-tester";
import prettier from "prettier";
import plugin from "babel-plugin-macros";

const macroFile = "'@c11/engine.macro'";
const config = require("./babelConfig.json");

pluginTester({
  plugin,
  babelOptions: { filename: __filename, ...config },
  formatResult: (result: any) => {
    return prettier.format(result, {
      parser: "babel",
    });
  },
  tests: {
    "should support Get": {
      code: `
        import { producer } from ${macroFile}
        const result: producer = ({
          a1 = Get.foo,
          a2 = Get[Prop.foo],
          a3 = Get[Arg.a2],
          a4 = Get.foo[Prop.bar.baz],
          a5 = Get.foo.bar[Arg.a2],
          a6 = {
            baz: Get.foo[Arg.a4].baz,
          },
          a7 = Get.foo[Arg.a5.baz],
          a8 = Get.foo[Param.prop],
          a9 = Get.foo[Param.prop].baz,
          a10 = Get.foo[Param.prop].baz[Prop.a2]
        }) => { }
      `,
      snapshot: true,
    },
  },
});
