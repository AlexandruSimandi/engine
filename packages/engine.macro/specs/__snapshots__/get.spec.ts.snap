// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros should support Get: should support Get 1`] = `

import { producer } from '@c11/engine.macro'
const result = producer((
  a1 = Get.foo,
  a2 = Get[Prop.foo],
  a3 = Get[Arg.a2],
  a4 = Get.foo[Prop.bar.baz],
  a5 = Get.foo.bar[Arg.a2],
  a6 = {
    baz: Get.foo[Arg.a4].baz,
  },
  a7 = Get.foo[Arg.a5.baz]
) => { })

      ↓ ↓ ↓ ↓ ↓ ↓

const result = {
  args: {
    type: "STRUCT",
    value: {
      a1: {
        type: "GET",
        path: [
          {
            type: "CONST",
            value: "foo"
          }
        ]
      },
      a2: {
        type: "GET",
        path: [
          {
            type: "EXTERNAL",
            path: ["foo"]
          }
        ]
      },
      a3: {
        type: "GET",
        path: [
          {
            type: "INTERNAL",
            path: ["a2"]
          }
        ]
      },
      a4: {
        type: "GET",
        path: [
          {
            type: "CONST",
            value: "foo"
          },
          {
            type: "EXTERNAL",
            path: ["bar", "baz"]
          }
        ]
      },
      a5: {
        type: "GET",
        path: [
          {
            type: "CONST",
            value: "foo"
          },
          {
            type: "CONST",
            value: "bar"
          },
          {
            type: "INTERNAL",
            path: ["a2"]
          }
        ]
      },
      a6: {
        type: "STRUCT",
        value: {
          baz: {
            type: "GET",
            path: [
              {
                type: "CONST",
                value: "foo"
              },
              {
                type: "INTERNAL",
                path: ["a4"]
              },
              {
                type: "CONST",
                value: "baz"
              }
            ]
          }
        }
      },
      a7: {
        type: "GET",
        path: [
          {
            type: "CONST",
            value: "foo"
          },
          {
            type: "INTERNAL",
            path: ["a5", "baz"]
          }
        ]
      }
    },
    meta: {
      order: ["a1", "a2", "a3", "a4", "a5", "a6", "a7"]
    }
  },
  fn: (a1, a2, a3, a4, a5, a6, a7) => {}
};


`;
